torch/csrc/jit/passes/pretty_print.cpp

```diff
@@ -0,0 +1,335 @@
+  const Graph& graph_;
+
+  // When printing a name if there is a conflict with an existing name in the
+  // graph, record the value -> new generated name mapping
+  std::unordered_map<const Value*, const Value*> aliases_;
+
+  // The Graph already tracks unique_names_, this is just for additional ones
+  // generated during printing
+  std::unordered_map<std::string, const Value*> generated_names_;
+
+  // Cache of value names
+  std::unordered_map<const Value*, std::string> value_names_;
+
+  template <class T>
+  void zipWith(
+      at::ArrayRef<T> list_a,
+      at::ArrayRef<T> list_b,
+      std::function<void(T, T)> action) const {
+    auto it_a = list_a.begin();
+    auto it_b = list_b.begin();
+
+    if (list_a.size() != list_b.size()) {
+      AT_ERROR("Pretty printer expected 2 lists of same size");
+    }
+
+    for (; it_a != list_a.end(); ++it_a, ++it_b) {
+      action(*it_a, *it_b);
+    }
+  }
+
+  std::ostream& printValueList(
+      std::ostream& out,
+      at::ArrayRef<const Value*> list) {
+    out << "(";
+    auto delimiter = "";
+    for (const auto* value : list) {
+      out << delimiter;
+      printValue(out, value);
+      delimiter = ", ";
+    }
+    out << ")";
+    return out;
+  }
+
+  void printAssignment(
+      std::ostream& out,
+      const Value* lhs,
+      const Value* rhs,
+      const size_t level) {
+    indent(out, level);
+    printValue(out, lhs);
+    out << " = ";
+    printValue(out, rhs);
+    out << "\n";
+  }
+
+  std::ostream& printIf(
+      std::ostream& out,
+      const Node* node,
+      const size_t level) {
+    indent(out, level);
+    out << "if ";
+    const auto if_block = node->blocks()[0];
+    const auto else_block = node->blocks()[1];
+    printValue(out, node->inputs()[0]);
+    out << ":"
+        << "\n";
+
+    // Print node contents
+    printBlock(out, if_block, level + 1);
+
+    // Print if block output
+    zipWith<const Value*>(
+        node->outputs(),
+        if_block->outputs(),
+        [&](const Value* node_output, const Value* return_input) {
+          printAssignment(out, node_output, return_input, level + 1);
+        });
+
+    indent(out, level);
+    out << "else:\n";
+    printBlock(out, else_block, level + 1);
+    zipWith<const Value*>(
+        node->outputs(),
+        else_block->outputs(),
+        [&](const Value* node_output, const Value* return_input) {
+          printAssignment(out, node_output, return_input, level + 1);
+        });
+
+    return out;
+  }
+
+  bool isValueUsedLater(const Value* val) const {
+    if (aliases_.find(val) != aliases_.end()) {
+      return true;
+    }
+    return val->uses().size() > 0;
+  }
+
+  std::ostream& printLoop(
+      std::ostream& out,
+      const Node* node,
+      const size_t level) {
+    const auto body_block = node->blocks()[0];
+    aliases_[body_block->inputs()[0]] = body_block->inputs()[0];
+
+    // Add temporaries for loop-carried dependencies
+    zipWith<const Value*>(
```

Comment:
Discussed in person: This should be possible to do by using only the `node->outputs()` names to represent the loop-carried dependencies and aliasing the `block->inputs()` names to be the `node->outputs()` name in the loop:

```
x1 = 6
y1 = 5
t = 4

x,y = Loop(x1, y1) 
         block(x2, y2)
         -> (t, x2)        
            

current:
            
  x2 = x1
  y2 = y1
  x = x2
  y = y2
  while ...:
    <body>
    x2 = t
    y2 = x2
    x = x2
    y = y2
  
  
x = x1
y = y1  
while ...:
    <body>
    x = t
    y = x2
```